### PR TITLE
mac stuff. also, better pip instructions

### PR DESCRIPTION
**Pull request by jamesyc**
added stuff for Mac.
Also, split Pip section out into it's own thing. Pip should be
installed before virtualenv. Mac OS X, installing homebrew to install
python means that Pip is installed before you even start THINKING about
installing virtualenv. The pip installation directions are unclear, so
I went digging into the
[history](https://pypi.python.org/pypi/setuptools/1.1.6#id104).

Originally, on the first day, after light and dark, God made python
setuptools; setuptools came with easy_install, which was a much easier
way to install python packages, and all was good. However, sin came
into the world and setuptools because old and not-quite-maintained.
Thus, Distribute came along, still with easy_install to install stuff.
Then from the tree of good and evil, came a better package manager in
the form of Pip.

So for a long time, installing distribute>pip was the best way to
install packages, as http://stackoverflow.com/a/5590714 shows. However,
distribute grew mature, and there came a time where the people who made
saw fit to merge it with the old setuptools, creating the new
setuptools. Pip followed suit, and from
https://github.com/pypa/pip/commit/0e342bc1297a4fde6d622d0f152161f3cbae9
162 we can see that it recommends new setuptools >=0.8.

Thus, the current best way to install the newest stuff is to $ wget
https://bitbucket.org/pypa/setuptools/raw/bootstrap/ez_setup.py | sudo
python && wget
https://raw.github.com/pypa/pip/master/contrib/get-pip.py | sudo
python. Aka install setuptools followed by pip.

Or perhaps, just $ sudo apt-get install python-setuptools python-pip

tl;dr pip comes before virtualenv, and figured out the proper way to install pip

### CODE REVIEW

**Comment by jamesyc** [comment]
whoops...

s/package manage/package manager/g

**Comment by sharadmv** [comment]
I don't think pip should be installed before virtualenv. virtualenv comes with pip, so it doesn't make sense to have an extra installation.

furthermore i think it's good practice to have pip sandboxed, because of python version.

the mac directions are good though. easy_install does come with macs. but I think i will keep the ordering (virtualenv then pip).

**Comment by jamesyc** [comment]
Pip is installed before virtualenv by default on macs. 

To clarify, easy_install is not installed by default on macs. Mac OS X comes with an ancient version of python (2.5?) or something, and that's it. In order to get macs up to date, you have to install the package manager "homebrew", and then run `$ brew install python`. Doing so installs setuptools (and thus easy_install) and Pip (all these _with_ the python install!) but NOT virtualenv. 

If you want to install pip after virtualenv on windows/linux, fine. But it's there on macs already, and it (installed outside of virtualenv) should still work fine with virtualenv. If you split it out that way, you're gonna need separate install directions (pip first for mac, pip after for win/nix). 

Also, I'd like to point out "virtualenv comes with pip" seems like pip is a prereq for virtualenv, not the other way around.

**Comment by sharadmv** [comment]
Sorry, wording issue. I meant to say, `pip` is included in `virtualenv`. I was under the impression I could just run `sudo easy_install virtualenv` on OSX and i would have virtualenv

**Comment by sharadmv** [comment]
Assuming `easy_install` is pre-installed or already installed

**Comment by jamesyc** [comment]
Nope. OS X doesn't even come with easy_install, or a modern version of python, for that matter. You need to install it with homebrew. 

AFTER you've installed homebrew and then install python, you can install virtualenv. However, by this point, you might as well as do "pip install virtualenv" rather than "easy_install virtualenv", since pip is already installed- and installing it this way, with pip, is cleaner.

From what I've seen in other places on the internet, like [here](http://stackoverflow.com/questions/5585875/what-is-the-official-preferred-way-to-install-pip-and-virtualenv-systemwide/5590714#5590714), the easy_install method is older (and deprecated?).

**Comment by sharadmv** [comment]
Ah I see. To maintain consistency with the other OSes in the guide, I think we should push to just get setuptools installed properly first, as it'll be confusing if we talk about `pip` outside the context of `virtualenv`. @richzeng how did you install virtualenv?

**Comment by jamesyc** [comment]
Eh, to clarify:

Mac OS X (comes with python 2.5, nothing else) -> homebrew -> `brew install python` (which simultaneously installs python 2.7, then setuptools (which contains easy_install) and pip). -> pip install virtualenv.

We can do something similar on ubuntu by doing, which keeps the most consistency:

Ubuntu -> `sudo apt-get install python` -> `sudo apt-get install python-pip` -> pip install virtualenv.

I don't think people will be confused if we tell them to install pip outside of virtualenv; hell, python itself is installed outside of virtualenv. We can just lump them together (which is pretty accurate anyways), and people can learn it that way. This method also has the benefit of not needing to tell people about setuptools/easy_install at all (it gets installed as a dependency of the python-pip package automagically), which removes a layer of abstraction. We can then just focus on pip as a python _package manager_, rather than easy_install as just a python _package installer_.

**Comment by sharadmv** [comment]
Oh no, i'm saying just do:

homebrew -> brew install python (which simultaneously installs setuptools (which contains easy_install), and pip). -> easy_install install virtualenv.

That should work right? If that's the case, i'd rather go with that.

**Comment by sharadmv** [comment]
The main reason being that on Windows, we need setuptools to install pip. So to stay consistent with that, i'd prefer that on OSX we do the same if we can

**Comment by jamesyc** [comment]
Ah, I see.

Alas, it's the other way around on mac- pip comes early enough in the chain, so you install virtualenv after pip.

We can install pip and setuptools at the same time, though, on windows. http://www.lfd.uci.edu/~gohlke/pythonlibs/#pip

Download both the setuptools and pip binary, and run them (it doesn't matter which order).

I decided to test this by doing it myself- i just fired up a windows 7 VM. By the way, it turns out that neither python nor pip are placed into the windows PATH when it installs. A pic can be seen here: http://puu.sh/4ygXI/d5c7ae611e.png

I think the best way to proceed on windows is:
windows (doesn't matter what version, x86/64) -> python 2.7 x86 -> put "C:\python27\" into PATH -> install setuptools and pip at the same time from http://www.lfd.uci.edu/~gohlke/pythonlibs/#pip -> put "C:\python27\scripts\" into the path -> install virtualenv with pip, which can now be called from cmd

**Comment by sharadmv** [comment]
I'm going to stick with setuptools. It works so far with our Windows guide. We're going to do the same for Macs (install virtualenv with setuptools)

**Comment by richzeng** [comment]
@jamesyc I've tried on three people's Macbooks who didn't have anything installed before they all had Python 2.7 and `easy_install` already installed. Are you sure about them starting with Python 2.5 and nothing else?

**Comment by jamesyc** [comment]
I'm not sure about it, I was going off stackoverflow posts, which had a guy running python 2.5. At any rate, I would highly recommend installing homebrew and using the python that comes from that instead of the default python for mac.

**Comment by jamesyc** [comment]
I just checked, the mac default install does contain easy_install, I was wrong. The mac os x default binaries are located in /usr/bin/, and installing homebrew etc puts stuff into /usr/local/bin, which should override the older binaries. 

I'd still like to point out that the binaries that come with OS X are pretty old. Ruby is some ancient version of 1.8.7. Python is 2.7.2, which i guess mountain lion has improved; however, apple is notorious for not ever updating those. Getting homebrew versions of these and being able to update all packages with a simple `brew update` would be much better.
